### PR TITLE
Fix configuration of Boost::python by trying different suffixes

### LINKED ISSUE #1333: Build fails on macOS
Following the updated instructions in the README:

> git submodule update --init --recursive
> mkdir build
> cd build
> cmake .. -DCMAKE_BUILD_TYPE=Release
> make -j$(nproc)
> sudo make install

`cmake` fails with the following output:

```
⇰ cmake .. -DCMAKE_BUILD_TYPE=Release
-- The CXX compiler identification is AppleClang 9.1.0.9020039
-- The C compiler identification is AppleClang 9.1.0.9020039
-- Check for working CXX compiler: /Library/Developer/CommandLineTools/usr/bin/c++
-- Check for working CXX compiler: /Library/Developer/CommandLineTools/usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Check for working C compiler: /Library/Developer/CommandLineTools/usr/bin/cc
-- Check for working C compiler: /Library/Developer/CommandLineTools/usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Configuring in release mode
-- Found PkgConfig: /usr/local/bin/pkg-config (found version "0.29.2")
-- Boost version: 1.67.0
-- Found the following Boost libraries:
--   date_time
--   filesystem
--   iostreams
--   regex
--   system
-- Found CURL: /usr/lib/libcurl.dylib (found version "7.54.0")
-- Found Lua: /usr/local/lib/liblua5.3.dylib (found version "5.3.4")
-- Found ZLIB: /usr/lib/libz.dylib (found version "1.2.11")
-- Performing Test LZ4_GOOD_VERSION
-- Performing Test LZ4_GOOD_VERSION - Success
-- Found LZ4: /usr/local/lib/liblz4.dylib
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Looking for pthread_create
-- Looking for pthread_create - found
-- Found Threads: TRUE
-- Checking for module 'libprime_server>=0.6.3'
--   Found libprime_server, version 0.6.3
-- Found Protobuf: /usr/local/lib/libprotobuf.dylib (found version "3.5.1")
-- Found SQLITE3: /usr/local/opt/sqlite3/lib/libsqlite3.dylib
-- Found SQLite3: /usr/local/opt/sqlite3/lib/libsqlite3.dylib
-- Looking for sqlite3_enable_load_extension in /usr/local/opt/sqlite3/lib/libsqlite3.dylib
-- Looking for sqlite3_enable_load_extension in /usr/local/opt/sqlite3/lib/libsqlite3.dylib - found
-- Found SpatiaLite: /usr/local/lib/libspatialite.dylib
-- Looking for gaiaStatisticsInvalidate in /usr/local/lib/libspatialite.dylib
-- Looking for gaiaStatisticsInvalidate in /usr/local/lib/libspatialite.dylib - found
-- Boost version: 1.67.0
-- Found the following Boost libraries:
--   program_options
-- Found GEOS: /usr/local/lib/libgeos.dylib;/usr/local/lib/libgeos_c.dylib
-- Found GEOS...
-- GEOS_INCLUDE_DIR=/usr/local/include
-- GEOS_LIBRARY=/usr/local/lib/libgeos.dylib;/usr/local/lib/libgeos_c.dylib
-- Found PythonInterp: /usr/local/bin/python (found version "2.7.15")
-- Found PythonLibs: /usr/lib/libpython2.7.dylib (found version "2.7.10")
-- Could NOT find Boost
-- Configuring done
CMake Error at /usr/local/Cellar/cmake/3.11.1/share/cmake/Modules/FindPythonLibs.cmake:288 (add_library):
  Target "python_valhalla" links to target "Boost::python" but the target was
  not found.  Perhaps a find_package() call is missing for an IMPORTED
  target, or an ALIAS target is missing?
Call Stack (most recent call first):
  CMakeLists.txt:162 (python_add_module)


-- Generating done
-- Build files have been written to: /Users/danielmactough/code/valhalla/build
```

### PR DESCRIPTION
**Pull request by oxidase**
# Issue

Fixes #1333 with a workaround that tries to find Boost::python, Boost::python2, Boost::python27

/cc @danpat @danpaz @danmactough @chaupow

## Tasklist

 - [x] review - you must request approval to merge any PR to master
 - [ ] Add #fixes with the issue number that this PR addresses
 - [ ] Generally use squash merge to rebase and clean comments before merging
 - [ ] Add to release notes in valhalla-docs
 - [ ] update relevant documentation if this is an API impacting change

## Requirements / Relations

 Link any requirements here. Other pull requests this PR is based on?

### CODE REVIEW

**Comment by danpat** [comment]
I managed to reproduce #1333 and this PR fixes things.

File: CMakeLists.txt [hunk 1]
```diff
@@ -157,14 +157,27 @@ add_subdirectory(test)
...
-    set(Boost_Python_lib "Boost::python")
-    find_package(Boost COMPONENTS python)
+    if(NOT Boost_PYTHON${PYTHON_VERSION_MAJOR}_FOUND)
+      set(Boost_Python_lib "Boost::python${PYTHON_VERSION_MAJOR}${PYTHON_VERSION_MINOR}")
+      set(_Boost_PYTHON${PYTHON_VERSION_MAJOR}${PYTHON_VERSION_MINOR}_HEADERS "boost/python.hpp")
+      find_package(Boost COMPONENTS python${PYTHON_VERSION_MAJOR}${PYTHON_VERSION_MINOR})
+      if(NOT Boost_PYTHON${PYTHON_VERSION_MAJOR}${PYTHON_VERSION_MINOR}_FOUND)
+        message(SEND_ERROR "Boost::python is not found")
+      endif()
+    endif()
```
Comment (danmactough) [inline CMakeLists.txt]: This is more comprehensive and less brittle than what I had done to get it working locally. 👍

**Comment by danpaz** [comment]
Worked for me after `brew install python3` :+1: thanks!

**Comment by danmactough** [comment]
@oxidase The build is working perfectly for me now (macOS)! 🎉 

However, the python module is getting installed to `/usr/local/lib/python2.7/site-packages/valhalla.so` instead of `/Library/Python/2.7/site-packages/valhalla.so`. I am able to resolve this by making the following change, but I'm not sure if this would break the linux build:

```diff
--- CMakeLists.txt
+++ CMakeLists.txt
@@ -175,7 +175,7 @@ if(ENABLE_PYTHON_BINDINGS)
   execute_process(
     COMMAND "${PYTHON_EXECUTABLE}" -c "
 from distutils import sysconfig as sc
-print(sc.get_python_lib(prefix='', plat_specific=True))"
+print(sc.get_python_lib(plat_specific=True))"
     OUTPUT_VARIABLE PYTHON_SITE_PACKAGES
     OUTPUT_STRIP_TRAILING_WHITESPACE)
   install(TARGETS python_valhalla
```

**Comment by oxidase** [comment]
@danmactough  `sc.get_python_lib(plat_specific=True)` returns absolute path `/usr/lib/python2.7/dist-packages` for python system packages  instead of relative one `lib/python2.7/dist-packages` that is combined with `CMAKE_INSTALL_PREFIX` (`/usr/local`).

The change will break installation on systems with FHS, but if for macOS the correct installation path is `/Library/Python/2.7/site-packages/` than the prefix of the `python_valhalla` install target for  must be  also updated.

**Comment by danpat** [comment]
On macOS, the `sys.path` for the system-provide Python binaries is:

```
$ python
Python 2.7.10 (default, Oct  6 2017, 22:29:07)
[GCC 4.2.1 Compatible Apple LLVM 9.0.0 (clang-900.0.31)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import sys
>>> print sys.path
['', '/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python27.zip', '/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7', '/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/plat-darwin', '/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/plat-mac', '/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/plat-mac/lib-scriptpackages', '/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/lib-tk', '/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/lib-old', '/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/lib-dynload', '/Library/Python/2.7/site-packages', '/System/Library/Frameworks/Python.framework/Versions/2.7/Extras/lib/python', '/System/Library/Frameworks/Python.framework/Versions/2.7/Extras/lib/python/PyObjC']
```

**Comment by danmactough** [comment]
Yeah, macOS is such a pain.

Here's what that script yields on macOS (system Python 2.7):

```python
⇰ python
Python 2.7.10 (default, Oct  6 2017, 22:29:07)
[GCC 4.2.1 Compatible Apple LLVM 9.0.0 (clang-900.0.31)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from distutils import sysconfig as sc
>>> sc.PREFIX
'/System/Library/Frameworks/Python.framework/Versions/2.7'
>>> sc.get_python_lib(prefix='', plat_specific=True)
'lib/python2.7/site-packages'
>>> sc.get_python_lib(plat_specific=True)
'/Library/Python/2.7/site-packages'
>>>
```

And the same with Homebrew Python 3:

```python
⇰ python3
Python 3.6.5 (default, Mar 30 2018, 06:41:53)
[GCC 4.2.1 Compatible Apple LLVM 9.0.0 (clang-900.0.39.2)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from distutils import sysconfig as sc
>>> sc.PREFIX
'/usr/local/Cellar/python/3.6.5/Frameworks/Python.framework/Versions/3.6'
>>> sc.get_python_lib(prefix='', plat_specific=True)
'lib/python3.6/site-packages'
>>> sc.get_python_lib(plat_specific=True)
'/usr/local/Cellar/python/3.6.5/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages'
>>>
```

**Comment by danmactough** [comment]
> sc.get_python_lib(plat_specific=True) returns absolute path /usr/lib/python2.7/dist-packages for python system packages instead of relative one lib/python2.7/dist-packages that is combined with CMAKE_INSTALL_PREFIX (/usr/local).

@oxidase I'm also seeing a problem with this trying to build python3 bindings on ubuntu (xenial). By combining that relative path with the `CMAKE_INSTALL_PREFIX` of `/usr/local`, the python module gets installed to a path that isn't in the python path: instead of `/usr/local/lib/python3/dist-packages` it should get installed to `/usr/lib/python3/dist-packages` ... or `/usr/local/lib/python3.5/dist-packages`.

Is it possible there's a more reliable way to construct these install paths? What's wrong with using the absolute path instead of smooshing together a prefix and a relative path, which is apparently error-prone?

**Comment by danmactough** [comment]
I know this is drifting a little off-the-point (fixing the macOS build), but here's what I have to do to get working python3 bindings built for Ubuntu xenial (copy/pasted from my Dockerfile), which leads me to wonder if we should be configuring cmake differently for these python bindings:

```sh
  cd /usr/local/src/valhalla && \
  mkdir build && \
  cd build && \
  ln -s /usr/lib/x86_64-linux-gnu/libboost_python-py35.so /usr/lib/x86_64-linux-gnu/libboost_python3.so && \
  ln -s /usr/local/lib/python3.5 /usr/local/lib/python3 && \
  cmake .. -DCMAKE_BUILD_TYPE=Release \
  -DENABLE_PYTHON_BINDINGS=On \
  -DPYTHON_EXECUTABLE=/usr/bin/python3 \
  -DPYTHON_INCLUDE_DIR=/usr/include/python3.5 \
  -DPYTHON_LIBRARY=/usr/lib/x86_64-linux-gnu/libpython3.5m.so && \
  make -j$(nproc) install
```

So that's:

- build needs `ln -s /usr/lib/x86_64-linux-gnu/libboost_python-py35.so /usr/lib/x86_64-linux-gnu/libboost_python3.so` (possibly [ubuntu-specific](https://stackoverflow.com/questions/39162622/boost-python3-missing-from-ubuntu-16-04))
- python3 module installation needs `ln -s /usr/local/lib/python3.5 /usr/local/lib/python3` (seeing this in debian jessie as well as ubuntu xenial)

**Comment by oxidase** [comment]
> Is it possible there's a more reliable way to construct these install paths? 

I've checked other repositories and used a part of [OpenCVDetectPython.cmake](https://github.com/opencv/opencv/blob/c729fbd652b1ebfbb205ac90c303160573f40dd1/cmake/OpenCVDetectPython.cmake#L133-L161) to detect python packages paths. 

> What's wrong with using the absolute path instead of smooshing together a prefix and a relative path, which is apparently error-prone?

There is nothing wrong with using absolute paths but installing python modules to `/usr/lib/python3/dist-packages` will break purpose of `/usr` directory  
[`Any information that is host-specific or varies with time is stored elsewhere.`](http://refspecs.linuxbase.org/FHS_3.0/fhs/ch04.html#purpose18)  Another issue with system paths may occur by making Valhalla packages  because a packaging framework expects that files are installed with `CMAKE_INSTALL_PREFIX`.

**Comment by chaupow** [comment]
I am going to test this for macOS:

test with
- [ ] no `boost-python`
- [ ] with `boost-python`
- [ ] python2 but not python3
- [ ] python3 but not python2
- [ ] python3 and python2

**Comment by danmactough** [comment]
@chaupow Are you using homebrew to provide any of those dependencies? I am using macOS-provided `python` (2.7) and homebrew installed `boost-python` (which only works for python2.7), `python3` and `boost-python3`. With that setup, I can run `cmake .. -DCMAKE_BUILD_TYPE=Release -DPYTHON_EXECUTABLE=$(which python3)` and then `make -j4 && sudo make install` and the installed python3 bindings work.

**Comment by chaupow** [comment]
Ohhhh @danmactough yea @oxidase also noticed that I was building it wrong.

I am using `cmake .. -DCMAKE_BUILD_TYPE=Release -DPYTHON_EXECUTABLE=$(which python3)` now and still get
```
>>> import valhalla; print(dir(valhalla))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ImportError: dynamic module does not define module export function (PyInit_valhalla)
```

But I guess if it runs for you on macOS, it must be my setup? I am using brew and have 
- `boost-python`
- `boost-python3`
- `python`
- `python@2`

installed

**Comment by danmactough** [comment]
> ImportError: dynamic module does not define module export function

Most likely explanation in this circumstance seems to be that you're including the python2 library in python3 or vice versa, in other words, [version mismatch](https://stackoverflow.com/a/37049403).

**Comment by oxidase** [comment]
@chaupow i have updated the branch and added a simple python test. Please can you check python bindings for version 2 and 3 as:
```
ccache -C
make clean
rm -rf CMakeCache.txt ; cmake -DPYTHON_EXECUTABLE=$(which python) .. -DCMAKE_INSTALL_PREFIX=/tmp/local
make -j4 run-python_valhalla
otool -L src/bindings/python/valhalla.so
src/bindings/python/valhalla.so:
	/usr/local/opt/python@2/Frameworks/Python.framework/Versions/2.7/Python (compatibility version 2.7.0, current version 2.7.0)
	/usr/local/opt/boost-python/lib/libboost_python-mt.dylib (compatibility version 0.0.0, current version 0.0.0)

rm -rf CMakeCache.txt ; cmake -DPYTHON_EXECUTABLE=$(which python3) .. -DCMAKE_INSTALL_PREFIX=/tmp/local
make -j4 run-python_valhalla
otool -L src/bindings/python/valhalla.so
src/bindings/python/valhalla.so:
	/usr/local/opt/python/Frameworks/Python.framework/Versions/3.6/Python (compatibility version 3.6.0, current version 3.6.0)
	/usr/local/opt/boost-python3/lib/libboost_python36-mt.dylib (compatibility version 0.0.0, current version 0.0.0)
```

@danmactough the updated install paths for `valhalla.so` are
```
/usr/local/lib/python2.7/site-packages/valhalla.so
/usr/local/lib/python3.6/site-packages/valhalla.so
```

**Comment by chaupow** [comment]
@oxidase thanks for all the effort 🙏 ❤️ 

here my `otool` output with `python`:
```
╰─ otool -L src/bindings/python/valhalla.so

src/bindings/python/valhalla.so:
	/System/Library/Frameworks/Python.framework/Versions/2.7/Python (compatibility version 2.7.0, current version 2.7.10)
	/usr/local/opt/boost-python/lib/libboost_python27-mt.dylib (compatibility version 0.0.0, current version 0.0.0)
	/usr/local/opt/protobuf/lib/libprotobuf.15.dylib (compatibility version 16.0.0, current version 16.1.0)
	/usr/local/opt/boost/lib/libboost_date_time-mt.dylib (compatibility version 0.0.0, current version 0.0.0)
	/usr/local/opt/boost/lib/libboost_filesystem-mt.dylib (compatibility version 0.0.0, current version 0.0.0)
	/usr/local/opt/boost/lib/libboost_iostreams-mt.dylib (compatibility version 0.0.0, current version 0.0.0)
	/usr/local/opt/boost/lib/libboost_system-mt.dylib (compatibility version 0.0.0, current version 0.0.0)
	/usr/lib/libcurl.4.dylib (compatibility version 7.0.0, current version 9.0.0)
	/usr/local/lib/libprime_server.0.dylib (compatibility version 1.0.0, current version 1.0.0)
	/usr/local/opt/boost/lib/libboost_regex-mt.dylib (compatibility version 0.0.0, current version 0.0.0)
	/usr/local/opt/lz4/lib/liblz4.1.dylib (compatibility version 1.0.0, current version 1.8.1)
	/usr/lib/libz.1.dylib (compatibility version 1.0.0, current version 1.2.11)
	/usr/local/opt/libspatialite/lib/libspatialite.7.dylib (compatibility version 9.0.0, current version 9.0.0)
	/usr/local/opt/sqlite/lib/libsqlite3.0.dylib (compatibility version 9.0.0, current version 9.6.0)
	/usr/local/opt/lua/lib/liblua.5.3.dylib (compatibility version 5.3.0, current version 5.3.4)
	/usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 400.9.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1252.50.4)
```

here my output from `otool` with `python3`
```
╰─ otool -L src/bindings/python/valhalla.so

src/bindings/python/valhalla.so:
	/usr/local/opt/python/Frameworks/Python.framework/Versions/3.6/Python (compatibility version 3.6.0, current version 3.6.0)
	/usr/local/opt/boost-python3/lib/libboost_python36-mt.dylib (compatibility version 0.0.0, current version 0.0.0)
	/usr/local/opt/protobuf/lib/libprotobuf.15.dylib (compatibility version 16.0.0, current version 16.1.0)
	/usr/local/opt/boost/lib/libboost_date_time-mt.dylib (compatibility version 0.0.0, current version 0.0.0)
	/usr/local/opt/boost/lib/libboost_filesystem-mt.dylib (compatibility version 0.0.0, current version 0.0.0)
	/usr/local/opt/boost/lib/libboost_iostreams-mt.dylib (compatibility version 0.0.0, current version 0.0.0)
	/usr/local/opt/boost/lib/libboost_system-mt.dylib (compatibility version 0.0.0, current version 0.0.0)
	/usr/lib/libcurl.4.dylib (compatibility version 7.0.0, current version 9.0.0)
	/usr/local/lib/libprime_server.0.dylib (compatibility version 1.0.0, current version 1.0.0)
	/usr/local/opt/boost/lib/libboost_regex-mt.dylib (compatibility version 0.0.0, current version 0.0.0)
	/usr/local/opt/lz4/lib/liblz4.1.dylib (compatibility version 1.0.0, current version 1.8.1)
	/usr/lib/libz.1.dylib (compatibility version 1.0.0, current version 1.2.11)
	/usr/local/opt/libspatialite/lib/libspatialite.7.dylib (compatibility version 9.0.0, current version 9.0.0)
	/usr/local/opt/sqlite/lib/libsqlite3.0.dylib (compatibility version 9.0.0, current version 9.6.0)
	/usr/local/opt/lua/lib/liblua.5.3.dylib (compatibility version 5.3.0, current version 5.3.4)
	/usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 400.9.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1252.50.4)
```


does this look good? 😁 Looks good to me.

**Pull request merged**